opm/simulators/linalg/bda/WellContributionsOCLContainer.cpp

```diff
@@ -0,0 +1,131 @@
+        Nb = Nb_;
+        dim = wellContribs.dim;
+        dim_wells = wellContribs.dim_wells;
+
+        if(!wellContribs.h_val_pointers_ocl.empty()){
+            num_std_wells = wellContribs.h_val_pointers_ocl.size() - 1;
+
+            s.Cnnzs = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * wellContribs.h_Cnnzs_ocl.size());
+            s.Dnnzs = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * wellContribs.h_Dnnzs_ocl.size());
+            s.Bnnzs = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * wellContribs.h_Bnnzs_ocl.size());
+            s.Ccols = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * wellContribs.h_Ccols_ocl.size());
+            s.Bcols = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * wellContribs.h_Bcols_ocl.size());
+            s.val_pointers = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(unsigned int) * wellContribs.h_val_pointers_ocl.size());
+            s.toOrder = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * Nb);
+        }
+        else{
+            num_std_wells = 0;
+        }
+    }
+
+    void WellContributionsOCLContainer::copy_to_gpu(Opm::WellContributions &wellContribs){
+        if(num_std_wells > 0){
+            toOrder.insert(toOrder.end(), wellContribs.toOrder, wellContribs.toOrder + Nb);
+
+            cl::Event event;
+            queue->enqueueWriteBuffer(s.Cnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Cnnzs_ocl.size(), wellContribs.h_Cnnzs_ocl.data(), nullptr, &event);
+            event.wait();
+            queue->enqueueWriteBuffer(s.Dnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Dnnzs_ocl.size(), wellContribs.h_Dnnzs_ocl.data(), nullptr, &event);
+            event.wait();
+            queue->enqueueWriteBuffer(s.Bnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Bnnzs_ocl.size(), wellContribs.h_Bnnzs_ocl.data(), nullptr, &event);
+            event.wait();
+            queue->enqueueWriteBuffer(s.Ccols, CL_TRUE, 0, sizeof(int) * wellContribs.h_Ccols_ocl.size(), wellContribs.h_Ccols_ocl.data(), nullptr, &event);
+            event.wait();
+            queue->enqueueWriteBuffer(s.Bcols, CL_TRUE, 0, sizeof(int) * wellContribs.h_Bcols_ocl.size(), wellContribs.h_Bcols_ocl.data(), nullptr, &event);
+            event.wait();
+            queue->enqueueWriteBuffer(s.val_pointers, CL_TRUE, 0, sizeof(unsigned int) * wellContribs.h_val_pointers_ocl.size(), wellContribs.h_val_pointers_ocl.data(), nullptr, &event);
+            event.wait();
+            queue->enqueueWriteBuffer(s.toOrder, CL_TRUE, 0, sizeof(int) * toOrder.size(), toOrder.data(), nullptr, &event);
+            event.wait();
+        }
+    }
+
+    void WellContributionsOCLContainer::update_on_gpu(Opm::WellContributions &wellContribs){
+        if(num_std_wells > 0){
+            cl::Event event;
+            queue->enqueueWriteBuffer(s.Cnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Cnnzs_ocl.size(), wellContribs.h_Cnnzs_ocl.data(), nullptr, &event);
```

Comment:
We are still writing to buffers that have been allocated during `init`. Can you comment on when that function is called and are you sure that the size is correct whenever this function is callled? As mentioned the size of the host buffers in wellContribs make change before each call to solve_system in the backend.